src/c++/perf_analyzer/perf_analyzer.cc

```diff
@@ -59,1887 +53,361 @@ SignalHandler(int signum)
+  if (params_->using_request_rate_range) {
+    if (params_->request_distribution == pa::Distribution::POISSON) {
       std::cout << "  Using poisson distribution on request generation"
                 << std::endl;
     } else {
       std::cout << "  Using uniform distribution on request generation"
                 << std::endl;
     }
   }
-  if (search_mode == pa::SearchMode::BINARY) {
+  if (params_->search_mode == pa::SearchMode::BINARY) {
     std::cout << "  Using Binary Search algorithm" << std::endl;
   }
-  if (async) {
+  if (params_->async) {
     std::cout << "  Using asynchronous calls for inference" << std::endl;
   } else {
     std::cout << "  Using synchronous calls for inference" << std::endl;
   }
-  if (parser->IsDecoupled()) {
+  if (parser_->IsDecoupled()) {
     std::cout << "  Detected decoupled model, using the first response for "
                  "measuring latency"
               << std::endl;
   }
 
-  if (percentile == -1) {
+  if (params_->percentile == -1) {
     std::cout << "  Stabilizing using average latency" << std::endl;
   } else {
-    std::cout << "  Stabilizing using p" << percentile << " latency"
+    std::cout << "  Stabilizing using p" << params_->percentile << " latency"
               << std::endl;
   }
   std::cout << std::endl;
+}
 
-  std::vector<pa::PerfStatus> summary;
-
-  if (using_custom_intervals) {
-    // Will be using user-provided time intervals, hence no control variable.
-    search_mode = pa::SearchMode::NONE;
-  }
-
-  mpi_driver->MPIBarrierWorld();
+void
+PerfAnalyzer::Profile()
+{
+  params_->mpi_driver->MPIBarrierWorld();
 
   cb::Error err;
-  if (target_concurrency) {
-    err = profiler->Profile<size_t>(
-        concurrency_range[SEARCH_RANGE::kSTART],
-        concurrency_range[SEARCH_RANGE::kEND],
-        concurrency_range[SEARCH_RANGE::kSTEP], search_mode, summary);
+  if (params_->targeting_concurrency()) {
+    err = profiler_->Profile<size_t>(
+        params_->concurrency_range.start, params_->concurrency_range.end,
+        params_->concurrency_range.step, params_->search_mode, summary_);
   } else {
-    err = profiler->Profile<double>(
-        request_rate_range[SEARCH_RANGE::kSTART],
-        request_rate_range[SEARCH_RANGE::kEND],
-        request_rate_range[SEARCH_RANGE::kSTEP], search_mode, summary);
+    err = profiler_->Profile<double>(
+        params_->request_rate_range[pa::SEARCH_RANGE::kSTART],
+        params_->request_rate_range[pa::SEARCH_RANGE::kEND],
+        params_->request_rate_range[pa::SEARCH_RANGE::kSTEP],
+        params_->search_mode, summary_);
   }
 
-  mpi_driver->MPIBarrierWorld();
+  params_->mpi_driver->MPIBarrierWorld();
 
   if (!err.IsOk()) {
     std::cerr << err;
     // In the case of early_exit, the thread does not return and continues to
     // report the summary
     if (!pa::early_exit) {
-      return err.Err();
+      throw pa::PerfAnalyzerException(err.Err());
```

Comment:
I like this clever way of incorporating C++ exceptions while still allowing return codes--nicely done 👍